src/refresh/vkpt/vertex_buffer.c

```diff
@@ -23,58 +23,395 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+		.dstAccessMask = VK_ACCESS_ACCELERATION_STRUCTURE_READ_BIT_KHR
+	};
+
+	vkCmdPipelineBarrier(cmd_buf, VK_PIPELINE_STAGE_ACCELERATION_STRUCTURE_BUILD_BIT_KHR,
+		VK_PIPELINE_STAGE_RAY_TRACING_SHADER_BIT_KHR, 0, 1,
+		&barrier, 0, 0, 0, 0);
+}
+
 VkResult
-vkpt_vertex_buffer_bsp_upload_staging()
+vkpt_vertex_buffer_upload_bsp_mesh(bsp_mesh_t* bsp_mesh)
 {
-	vkWaitForFences(qvk.device, 1, &qvk.fence_vertex_sync, VK_TRUE, ~((uint64_t)0));
-	vkResetFences(qvk.device, 1, &qvk.fence_vertex_sync);
+	assert(bsp_mesh);
+
+	vkDeviceWaitIdle(qvk.device);
+
+	size_t vbo_size = bsp_mesh->num_primitives * sizeof(VboPrimitive);
+	bsp_mesh->vertex_data_offset = vbo_size;
+	vbo_size += bsp_mesh->num_primitives * sizeof(mat3);
+	size_t staging_size = vbo_size;
+
+	suballocate_model_blas_memory(&bsp_mesh->geom_opaque,      &vbo_size, "bsp:opaque");
+	suballocate_model_blas_memory(&bsp_mesh->geom_transparent, &vbo_size, "bsp:transparent");
+	suballocate_model_blas_memory(&bsp_mesh->geom_masked,      &vbo_size, "bsp:masked");
+	suballocate_model_blas_memory(&bsp_mesh->geom_sky,         &vbo_size, "bsp:sky");
+	suballocate_model_blas_memory(&bsp_mesh->geom_custom_sky,  &vbo_size, "bsp:custom_sky");
+
+	for (int i = 0; i < bsp_mesh->num_models; i++)
+	{
+		bsp_model_t* model = bsp_mesh->models + i;
+
+		char name[MAX_QPATH];
+		Q_snprintf(name, sizeof(name), "bsp:models[%d]", i);
+		
+		suballocate_model_blas_memory(&model->geometry, &vbo_size, name);
+	}
+	
+	VkResult res = buffer_create(&qvk.buf_world, vbo_size,
+		VK_BUFFER_USAGE_TRANSFER_DST_BIT |
+		VK_BUFFER_USAGE_STORAGE_BUFFER_BIT |
+		VK_BUFFER_USAGE_VERTEX_BUFFER_BIT |
+		VK_BUFFER_USAGE_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT_KHR |
+		VK_BUFFER_USAGE_ACCELERATION_STRUCTURE_STORAGE_BIT_KHR |
+		VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT,
+		VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT);
+
+	if (res != VK_SUCCESS) return res;
+
+	BufferResource_t staging_buffer;
+
+	res = buffer_create(&staging_buffer, staging_size,
+		VK_BUFFER_USAGE_TRANSFER_SRC_BIT,
+		VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT |
+		VK_MEMORY_PROPERTY_HOST_COHERENT_BIT);
+
+	if (res != VK_SUCCESS) return res;
+
+	create_model_blas(&bsp_mesh->geom_opaque,      qvk.buf_world.buffer);
+	create_model_blas(&bsp_mesh->geom_transparent, qvk.buf_world.buffer);
+	create_model_blas(&bsp_mesh->geom_masked,      qvk.buf_world.buffer);
+	create_model_blas(&bsp_mesh->geom_sky,         qvk.buf_world.buffer);
+	create_model_blas(&bsp_mesh->geom_custom_sky,  qvk.buf_world.buffer);
+
+	for (int i = 0; i < bsp_mesh->num_models; i++)
+	{
+		bsp_model_t* model = bsp_mesh->models + i;
+		create_model_blas(&model->geometry, qvk.buf_world.buffer);
+	}
+
+	uint8_t* staging_data = buffer_map(&staging_buffer);
+	memcpy(staging_data, bsp_mesh->primitives, bsp_mesh->num_primitives * sizeof(VboPrimitive));
+
+	mat3* positions = (mat3*)(staging_data + bsp_mesh->vertex_data_offset);  // NOLINT(clang-diagnostic-cast-align)
```

Comment:
IMO, since this is dealing with triangles, using a type (name) other than `mat3` could be clearer.